core/types/transaction_conditional.go

```diff
@@ -0,0 +1,130 @@
+	return nil
+}
+
+// MarshalJSON will serialize the KnownAccount into JSON bytes.
+func (ka *KnownAccount) MarshalJSON() ([]byte, error) {
+	if ka.StorageRoot != nil {
+		return json.Marshal(ka.StorageRoot)
+	}
+	return json.Marshal(ka.StorageSlots)
+}
+
+// Root will return the storage root and true when the user prefers
+// execution against an account's storage root, otherwise it will
+// return false.
+func (ka *KnownAccount) Root() (common.Hash, bool) {
+	if ka.StorageRoot == nil {
+		return common.Hash{}, false
+	}
+	return *ka.StorageRoot, true
+}
+
+// Slots will return the storage slots and true when the user prefers
+// execution against an account's particular storage slots, StorageRoot == nil,
+// otherwise it will return false.
+func (ka *KnownAccount) Slots() (map[common.Hash]common.Hash, bool) {
+	if ka.StorageRoot != nil {
+		return ka.StorageSlots, false
+	}
+	return ka.StorageSlots, true
+}
+
+//go:generate go run github.com/fjl/gencodec -type TransactionConditional -field-override transactionConditionalMarshalling -out gen_transaction_conditional_json.go
+
+// TransactionConditional represents the preconditions that determine the
+// inclusion of the transaction, enforced out-of-protocol by the sequencer.
+type TransactionConditional struct {
+	// KnownAccounts represents account prestate conditions
+	KnownAccounts KnownAccounts `json:"knownAccounts"`
+
+	// Header state conditionals (inclusive ranges)
+	BlockNumberMin *big.Int `json:"blockNumberMin,omitempty"`
+	BlockNumberMax *big.Int `json:"blockNumberMax,omitempty"`
+	TimestampMin   *uint64  `json:"timestampMin,omitempty"`
+	TimestampMax   *uint64  `json:"timestampMax,omitempty"`
+
+	// Internal fields used for metrics and inclusion tracking
+	SubmissionTime time.Time    `json:"-"`
+	Rejected       *atomic.Bool `json:"-"`
+}
+
+// field type overrides for gencodec
+type transactionConditionalMarshalling struct {
+	BlockNumberMax *hexutil.Big
+	BlockNumberMin *hexutil.Big
+	TimestampMin   *hexutil.Uint64
+	TimestampMax   *hexutil.Uint64
+}
+
+// Validate will perform sanity checks on the preconditions. This does not check the aggregate cost of the preconditions.
+func (cond *TransactionConditional) Validate() error {
+	if cond.BlockNumberMin != nil && cond.BlockNumberMax != nil && cond.BlockNumberMin.Cmp(cond.BlockNumberMax) > 0 {
+		return fmt.Errorf("block number minimum constraint must be less than the maximum")
+	}
+	if cond.TimestampMin != nil && cond.TimestampMax != nil && *cond.TimestampMin > *cond.TimestampMax {
+		return fmt.Errorf("timestamp minimum constraint must be less than the maximum")
+	}
+	return nil
+}
+
+// Cost computes the aggregate cost of the preconditions; total number of storage lookups required
+func (cond *TransactionConditional) Cost() int {
+	cost := 0
+	for _, account := range cond.KnownAccounts {
+		if _, isRoot := account.Root(); isRoot {
+			cost += 1
+		}
+		if slots, isSlots := account.Slots(); isSlots {
+			cost += len(slots)
+		}
+	}
```

Comment:
Is this cost function standardized somewhere? I think it makes sense to always add `1` for an account, even if the account is empty, or if there are slots but no root. Since we end up doing a little bit of work per known account entry. E.g. we wouldn't want a super-long list of accounts that we all count as 0 cost, because the accounts-list still fills up memory.

Reply:
I do not believe it is standardized, it was one thing that I thought was lacking in the original spec

Reply:
adding +1 as a default per account sounds good to me. Can adjust later on